File: currency_converter.swift
// currency_converter.swift
import Foundation

let CACHE_TTL: TimeInterval = 3600 // 1 hour
var cachedRates: [String: Double]?
var cacheTimestamp: Date?

func fetchRates() -> [String: Double]? {
    let url = URL(string: "https://api.exchangerate.host/latest?base=USD")!
    let semaphore = DispatchSemaphore(value: 0)
    var result: [String: Double]?
    let task = URLSession.shared.dataTask(with: url) { data, response, error in
        defer { semaphore.signal() }
        guard let data = data, error == nil else { return }
        do {
            let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
            if let success = json?["success"] as? Bool, success == false {
                return
            }
            if let rates = json?["rates"] as? [String: Double] {
                result = rates
            }
        } catch {
            // ignore
        }
    }
    task.resume()
    semaphore.wait()
    return result
}

func getRates() -> [String: Double] {
    let now = Date()
    if cachedRates == nil || cacheTimestamp == nil || now.timeIntervalSince(cacheTimestamp!) > CACHE_TTL {
        print("🔄 Fetching fresh exchange rates...")
        if let newRates = fetchRates() {
            cachedRates = newRates
            cacheTimestamp = now
        } else {
            if cachedRates == nil {
                print("❌ Failed to fetch rates and no cache. Exiting.")
                exit(1)
            }
            print("⚠️  Using cached rates (stale).")
        }
    }
    return cachedRates!
}

func convert(amount: Double, from fromCur: String, to toCur: String) throws -> Double {
    let rates = getRates()
    var usdAmount: Double
    if fromCur == "USD" {
        usdAmount = amount
    } else {
        guard let fromRate = rates[fromCur] else {
            throw NSError(domain: "", code: 1, userInfo: [NSLocalizedDescriptionKey: "Currency \(fromCur) not supported."])
        }
        usdAmount = amount / fromRate
    }
    if toCur == "USD" {
        return usdAmount
    }
    guard let toRate = rates[toCur] else {
        throw NSError(domain: "", code: 1, userInfo: [NSLocalizedDescriptionKey: "Currency \(toCur) not supported."])
    }
    return usdAmount * toRate
}

func showCurrencies() {
    let rates = getRates()
    print("\nAvailable currencies (rates vs USD):")
    for (curr, rate) in rates.sorted(by: { $0.key < $1.key }) {
        print("  \(curr): \(String(format: "%.4f", rate))")
    }
}

func main() {
    print("=== Currency Converter with Caching ===")
    while true {
        print("\n1. Convert")
        print("2. Show available currencies")
        print("3. Refresh cache")
        print("4. Exit")
        print("Choose: ", terminator: "")
        guard let choice = readLine()?.trimmingCharacters(in: .whitespacesAndNewlines) else { continue }
        switch choice {
        case "1":
            print("Enter amount: ", terminator: "")
            guard let amountStr = readLine(), let amount = Double(amountStr), amount >= 0 else {
                print("❌ Invalid amount.")
                continue
            }
            print("From currency (e.g., USD): ", terminator: "")
            let fromCur = readLine()?.trimmingCharacters(in: .whitespaces).uppercased() ?? ""
            print("To currency (e.g., EUR): ", terminator: "")
            let toCur = readLine()?.trimmingCharacters(in: .whitespaces).uppercased() ?? ""
            do {
                let result = try convert(amount: amount, from: fromCur, to: toCur)
                print(String(format: "%.2f %@ = %.2f %@", amount, fromCur, result, toCur))
            } catch {
                print("❌ \(error.localizedDescription)")
            }
        case "2":
            showCurrencies()
        case "3":
            cachedRates = nil
            cacheTimestamp = nil
            print("Cache cleared. Will refresh on next operation.")
        case "4":
            print("Goodbye!")
            return
        default:
            print("Invalid choice.")
        }
    }
}

main()

File: CurrencyConverter.java
// CurrencyConverter.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.time.Duration;
import java.time.Instant;
import java.util.*;
import java.util.stream.Collectors;

public class CurrencyConverter {
    private static Map<String, Double> rates = null;
    private static Instant timestamp = null;
    private static final Duration CACHE_TTL = Duration.ofHours(1);

    public static Map<String, Double> fetchRates() throws Exception {
        String urlStr = "https://api.exchangerate.host/latest?base=USD";
        URL url = new URL(urlStr);
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("GET");
        conn.setConnectTimeout(5000);
        conn.setReadTimeout(5000);

        if (conn.getResponseCode() != 200) {
            throw new RuntimeException("HTTP error: " + conn.getResponseCode());
        }

        try (BufferedReader br = new BufferedReader(new InputStreamReader(conn.getInputStream()))) {
            String json = br.lines().collect(Collectors.joining());
            // Naive JSON parsing – assuming format: {"success":true,"rates":{"USD":1.0,...}}
            // For production, use a JSON library (Jackson/Gson). We'll do simple parsing.
            // We'll parse manually for brevity.
            if (!json.contains("\"success\":true")) {
                throw new RuntimeException("API returned success=false");
            }
            // Extract rates object
            int ratesStart = json.indexOf("\"rates\":{") + 8;
            int ratesEnd = json.lastIndexOf("}");
            String ratesJson = json.substring(ratesStart, ratesEnd);
            // Split by comma, but careful: values contain commas? Not in numbers.
            Map<String, Double> result = new HashMap<>();
            // Remove leading/trailing braces
            ratesJson = ratesJson.trim();
            if (ratesJson.startsWith("{")) ratesJson = ratesJson.substring(1);
            if (ratesJson.endsWith("}")) ratesJson = ratesJson.substring(0, ratesJson.length()-1);
            String[] pairs = ratesJson.split(",");
            for (String pair : pairs) {
                String[] kv = pair.split(":");
                if (kv.length == 2) {
                    String key = kv[0].trim().replaceAll("\"", "");
                    double value = Double.parseDouble(kv[1].trim());
                    result.put(key, value);
                }
            }
            return result;
        }
    }

    public static Map<String, Double> getRates() {
        Instant now = Instant.now();
        if (rates == null || timestamp == null || Duration.between(timestamp, now).compareTo(CACHE_TTL) > 0) {
            System.out.println("🔄 Fetching fresh exchange rates...");
            try {
                rates = fetchRates();
                timestamp = now;
            } catch (Exception e) {
                if (rates == null) {
                    System.err.println("❌ Failed to fetch rates and no cache: " + e.getMessage());
                    System.exit(1);
                }
                System.out.println("⚠️  Using cached rates (stale).");
            }
        }
        return rates;
    }

    public static double convert(double amount, String fromCur, String toCur) throws Exception {
        Map<String, Double> rates = getRates();
        double usdAmount;
        if (fromCur.equals("USD")) {
            usdAmount = amount;
        } else {
            Double fromRate = rates.get(fromCur);
            if (fromRate == null) throw new Exception("Currency " + fromCur + " not supported.");
            usdAmount = amount / fromRate;
        }
        if (toCur.equals("USD")) return usdAmount;
        Double toRate = rates.get(toCur);
        if (toRate == null) throw new Exception("Currency " + toCur + " not supported.");
        return usdAmount * toRate;
    }

    public static void showCurrencies() {
        Map<String, Double> rates = getRates();
        System.out.println("\nAvailable currencies (rates vs USD):");
        rates.entrySet().stream().sorted(Map.Entry.comparingByKey())
                .forEach(e -> System.out.printf("  %s: %.4f%n", e.getKey(), e.getValue()));
    }

    public static void main(String[] args) throws Exception {
        Scanner scanner = new Scanner(System.in);
        System.out.println("=== Currency Converter with Caching ===");
        while (true) {
            System.out.println("\n1. Convert");
            System.out.println("2. Show available currencies");
            System.out.println("3. Refresh cache");
            System.out.println("4. Exit");
            System.out.print("Choose: ");
            String choice = scanner.nextLine().trim();
            switch (choice) {
                case "1":
                    System.out.print("Enter amount: ");
                    double amount;
                    try {
                        amount = Double.parseDouble(scanner.nextLine().trim());
                    } catch (NumberFormatException e) {
                        System.out.println("❌ Invalid amount.");
                        continue;
                    }
                    if (amount < 0) {
                        System.out.println("❌ Amount must be positive.");
                        continue;
                    }
                    System.out.print("From currency (e.g., USD): ");
                    String fromCur = scanner.nextLine().trim().toUpperCase();
                    System.out.print("To currency (e.g., EUR): ");
                    String toCur = scanner.nextLine().trim().toUpperCase();
                    try {
                        double result = convert(amount, fromCur, toCur);
                        System.out.printf("%.2f %s = %.2f %s%n", amount, fromCur, result, toCur);
                    } catch (Exception e) {
                        System.out.println("❌ " + e.getMessage());
                    }
                    break;
                case "2":
                    showCurrencies();
                    break;
                case "3":
                    rates = null;
                    timestamp = null;
                    System.out.println("Cache cleared. Will refresh on next operation.");
                    break;
                case "4":
                    System.out.println("Goodbye!");
                    scanner.close();
                    return;
                default:
                    System.out.println("Invalid choice.");
            }
        }
    }
}
